### PR TITLE
[GCU] Handling type1 lists

### LINKED ISSUE #2034: [GCU] Update path to xpath conversion logic to most recent sonic YANG updates Supporting `Type_1_list_maps_model
#### Description

sonic_yang_mgmt follows these [guidelines](https://github.com/Azure/SONiC/blob/master/doc/mgmt/SONiC_YANG_Model_Guidelines.md) in order to convert from `config_db` to `config_yang` and vice versa.
in GCU we rely on the same [guidelines](https://github.com/Azure/SONiC/blob/master/doc/mgmt/SONiC_YANG_Model_Guidelines.md) to convert from `config_db path` to `config_yang xpath` and vice versa.

There were modifications to how sonic_yang_mgmt converts data recently, and these changes are not supported by gcu, the changes are the following:
- Supporting `Type_1_list_maps_model ` https://github.com/Azure/sonic-buildimage/pull/9116

Please make sure to add tests for CABLE_LENGTH and other tables that are affected by this change.

## Steps to reproduce the issue
1.
2.
3.

#### Output of `show version`

```
(paste your output here)
```

### PR DESCRIPTION
**Pull request by ghooo**
#### What I did
Fixes #2034

Which lists where handled before in ConfigDb to YANG conversion?
- Dictionary of key to Object e.g.
```
"TACPLUS_SERVER": {
    "1.1.1.1": {            <= Key: Object (i.e. {...})
        "priority": "1", 
        "tcp_port": "49"
    }, 
    "1.1.1.2": {
        "priority": "1", 
        "tcp_port": "49"
    },
    "1.1.1.3": {
        "priority": "1", 
        "tcp_port": "49"
    } 
}
```
- List of values
```
"VLAN": {
    "Vlan1000": {
        "vlanid": "1000",
        "dhcp_servers": [    <= Values
            "192.0.0.1",
            "192.0.0.2",
            "192.0.0.3",
            "192.0.0.4"
        ]
    }
}
```
- But there is no handling of Dictionary of Key to Value
```
"DOT1P_TO_TC_MAP": {
    "Dot1p_to_tc_map1": {    <= Key: Value
            "1": "1",
            "2": "2",
            "3": "1",
            "4": "2"
        }
    },
```

Refer to https://github.com/Azure/sonic-buildimage/pull/7375 for more info about how Type 1 list ConfigDb is getting converted to SonicYang.

After checking how type1 is handled during ConfigDb to SonicYang conversion. Check the unit-tests here for converting ConfigDb Path to SonicYang xpath.

Also added CABLE_LENGTH to ADD_RACK and REMOVE_RACK tests.

#### How I did it

When handling a list, check if it is of type1. If that's the case call type1 list handling.

#### How to verify it
unit-test

### CODE REVIEW

File: generic_config_updater/gu_common.py [hunk 1]
```diff
@@ -711,10 +735,63 @@ def _get_path_tokens_from_list(self, model, token_index, xpath_tokens, config):
...
+        # Example where VLAN_MEMBER_LIST has no specific key/value:
+        #   xpath: /sonic-dot1p-tc-map:sonic-dot1p-tc-map/DOT1P_TO_TC_MAP/DOT1P_TO_TC_MAP_LIST[name='Dot1p_to_tc_map1']/DOT1P_TO_TC_MAP
+        #   path: /DOT1P_TO_TC_MAP/Dot1p_to_tc_map1
+        if not(key_dict):
+            return []
+
+        if len(key_dict) > 1:
+            raise GenericConfigUpdaterError(f"Type 1 inner list should have only 1 key in xpath, {len(key_dict)} specified. Key dictionary: {key_dict}")
+
+        keyName = list(key_dict.keys())[0]
```
Comment (qiluo-msft) [inline generic_config_updater/gu_common.py:772]: >key_dict.keys() [](http://example.com/codeflow?start=23&length=15)

`next(iter(key_dict.keys()))` is more efficient.

File: generic_config_updater/gu_common.py [hunk 2]
```diff
@@ -711,10 +735,63 @@ def _get_path_tokens_from_list(self, model, token_index, xpath_tokens, config):
...
+
+        # Checking if the next_token is actually a child leaf of the inner type 1 list, for which case
+        # just ignore the token, and return the already created ConfigDb path pointing to the whole object
+        # Example where the leaf specified is the key:
+        #   xpath: /sonic-dot1p-tc-map:sonic-dot1p-tc-map/DOT1P_TO_TC_MAP/DOT1P_TO_TC_MAP_LIST[name='Dot1p_to_tc_map1']/DOT1P_TO_TC_MAP[dot1p='2']/dot1p
+        #   path: /DOT1P_TO_TC_MAP/Dot1p_to_tc_map1/2
+        # Example where the leaf specified is not the key:
+        #   xpath: /sonic-dot1p-tc-map:sonic-dot1p-tc-map/DOT1P_TO_TC_MAP/DOT1P_TO_TC_MAP_LIST[name='Dot1p_to_tc_map1']/DOT1P_TO_TC_MAP[dot1p='2']/tc
+        #   path: /DOT1P_TO_TC_MAP/Dot1p_to_tc_map1/2
+        next_token = xpath_tokens[token_index+1]
```
Comment (qiluo-msft) [inline generic_config_updater/gu_common.py:788]: >token_index [](http://example.com/codeflow?start=34&length=11)

Protect for index out of range.

Comment (ghooo) [reply generic_config_updater/gu_common.py:788]: it is protected in line 777

Comment (ghooo) [reply generic_config_updater/gu_common.py:788]: let me add a comment to line 777 for clarification, and to indicate that it is the end of the xpath_tokens

File: generic_config_updater/gu_common.py (hunk 1 above)
Comment (ghooo) [reply generic_config_updater/gu_common.py:772]: sure

File: generic_config_updater/gu_common.py [hunk 3]
```diff
@@ -711,10 +735,63 @@ def _get_path_tokens_from_list(self, model, token_index, xpath_tokens, config):
         if next_token in key_dict:
             return path_tokens
 
+        list_name = model['@name']
+        type_1_list_model = self._get_type_1_list_model(model, list_name)
```
[the remark this answers is not in the document]
Comment (ghooo) [reply generic_config_updater/gu_common.py:739]: good idea

[the remark this answers is not in the document]
Comment (ghooo) [reply generic_config_updater/gu_common.py]: Updated

File: generic_config_updater/gu_common.py (hunk 1 above)
Comment (ghooo) [reply generic_config_updater/gu_common.py]: Updated

File: generic_config_updater/gu_common.py (hunk 2 above)
Comment (ghooo) [reply generic_config_updater/gu_common.py:793]: Updated

Comment (qiluo-msft) [reply generic_config_updater/gu_common.py:792]: How about `token_index+1 > len(xpath_tokens)`?

**Comment by qiluo-msft** [comment]
Please resolve conflict.

**Comment by ghooo** [comment]
> Please resolve conflict.

Resolved

**Review by qiluo-msft — APPROVED** [review]

**Pull request merged**